### ISSUE TITLE
Differences in model output between Ray and Local backends.

### ISSUE THREAD

**Issue by geoffreyangus** [ludwig-ai/ludwig#2054]
**Describe the bug**
Given some trained model with weights M and some dataset D, M evaluated on D using a Ray backend and M evaluated on D using a Local backend produces different numerical outputs.

**To Reproduce**

1. Add print statements to look at the predictions from the model evaluation calls in `tests.integration_tests.utils.train_with_backend`.
2. Run a test that calls that function, e.g. `tests/integration_tests/test_ray.py::test_ray_image`.

**Expected behavior**
I expect evaluation calls on the same model with the same dataset to produces the same numerical outputs regardless of the backend.

**Environment (please complete the following information):**

- OS: macOS
- Version: 12.3
- Python version: 3.8
- Ludwig version: 0.5

**Additional context**

We currently workaround this by having some degree of tolerance when comparing the models in unit tests.

**Comment by geoffreyangus** [comment]
Update: I am seeing unit tests fail because AUC values at model evaluation time are different between Ray and Local backends.

The test proceeds as following:

1. A CSV of 100 samples with 10% NaNs in each column is generated as dataset `D`. 
2. A model `M_ray` is instantiated with a Ray backend.
2. `M_ray` is trained on `D`.
3. `M_ray` is evaluated on `D`, creating `metric_ray`.
4. `M_ray` is saved, then reloaded with a Local backend as `M_local`.
5. `M_local` is evaluated on `D`, creating `metric_local`.
6. `metric_ray` is compared with `metric_local`.

Sometimes the test passes and sometimes it fails. I confirmed the following:

1. The data looks identical between the two backends. This is right before the forward pass.
2. The output probabilities look identical between the two backends. This is a change from what I initially thought– the reason being that I was not considering that batches were being split between two Ray processes.
3. Tests that were passing were processing 90 samples. This is expected because drop_rows should be removing 10% of samples (the ones with NaNs).
3. Tests that were failing were processing 100 samples. This is unexpected for the same reason as above.

## Inconsistencies in Dask type inference

It seems that the issue is related to WHERE exactly the first NaN is located in the binary feature. Such an explanation is supported by the [Dask docs](https://docs.dask.org/en/stable/generated/dask.dataframe.read_csv.html): 

> Dask dataframe tries to infer the dtype of each column by reading a sample from the start of the file (or of the first file if it’s a glob). Usually this works fine, but if the dtype is different later in the file (or in other files) this can cause issues. For example, if all the rows in the sample had integer dtypes, but later on there was a NaN, then this would error at compute time.

First, inspect the dataframe instantiated using `pd.read_csv`. The binary feature is of type `object` because of the presence of NaNs. Note that in this failing test, the first NaN of the binary column appears in the row with index `10`.


```python
pd.read_csv('failing_dataset.csv').head(15)
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>audio_58299</th>
      <th>binary_EB37B</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>3</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>4</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>5</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>6</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>7</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>8</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>9</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>10</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>11</th>
      <td>NaN</td>
      <td>True</td>
    </tr>
    <tr>
      <th>12</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>13</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>14</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
</div>




```python
pd.read_csv('failing_datasetdataset.csv')['binary_EB37B'].dtype
```




    dtype('O')



When we read the same dataset using Dask, the NaN at index 10 disappears and is replaced with a `True`. Further inspection shows that Dask inferred the dtype of this column to be `bool`, which does not allow the presence of NaNs.


```python
dd.read_csv('failing_dataset.csv').compute().head(15)
```




<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>audio_58299</th>
      <th>binary_EB37B</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>1</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>2</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>3</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>4</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>5</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>6</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>7</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>8</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>9</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>10</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>True</td>
    </tr>
    <tr>
      <th>11</th>
      <td>NaN</td>
      <td>True</td>
    </tr>
    <tr>
      <th>12</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>13</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
    <tr>
      <th>14</th>
      <td>/var/folders/23/8ddc07cj3knbkldhnfv2cmbr0000gn...</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
</div>




```python
dd.read_csv('/Users/geoffreyangus/Downloads/unittest/dataset.csv').compute()['binary_EB37B'].dtype
```




    dtype('bool')

**Comment by geoffreyangus** [comment]
Fixed in #2058.
